MoltenVK/MoltenVK/GPUObjects/MVKSync.h

```diff
@@ -636,3 +636,56 @@ class MVKMetalCompiler : public MVKBaseObject {
 	std::string _compilerType = "Unknown";
 	MVKPerformanceTracker* _pPerformanceTracker = nullptr;
 };
+
+#pragma mark -
+#pragma mark MVKDeferredOperation
+
+/** Defines the function pointer for each dependent function*/
+union MVKDeferredOperationFunctionPointer
+{
+    // Empty until deferred functions from other extensions have been defined
+    // Planning to use std::functions
+};
+
+/** Manages what function is being deferred*/
+enum MVKDeferredOperationFunctionType
+{
+    // Empty until deferred functions from other extensions have been defined
+};
+
+class MVKDeferredOperation : public MVKVulkanAPIDeviceObject {
+public:
+    /** Returns the Vulkan type of this object. */
+    VkObjectType getVkObjectType() override { return VK_OBJECT_TYPE_DEFERRED_OPERATION_KHR; }
+    
+    /** Returns the debug report object type of this object. */
+    VkDebugReportObjectTypeEXT getVkDebugReportObjectType() override { return VK_DEBUG_REPORT_OBJECT_TYPE_UNKNOWN_EXT; }
+    
+    /** Joins the deferred operation */
+    VkResult join();
+    
+    /** Gets the max number of threads that can execute the deferred operation concurrently*/
+    uint32_t getMaxConcurrency() { return 1; } // Perhaps the number of CPU cores
+    
+    /** Gets the result of the execution of the deferred operation */
+    VkResult getResult() { return operationResult; }
+    
+    /** Sets all the variables needed for a deferred operation, however should never be called manually and only from other functions that take deferred operations*/
+    void deferOperation(MVKDeferredOperationFunctionPointer pointer, MVKDeferredOperationFunctionType type, std::vector<void*> parameters);
+#pragma mark Construction
+    MVKDeferredOperation(MVKDevice* device) : MVKVulkanAPIDeviceObject(device) {}
+protected:
+    /** Stores the result of the operation*/
+    VkResult operationResult = VK_SUCCESS;
+    
+    /** Stores a pointer to the function*/
+    MVKDeferredOperationFunctionPointer functionPointer;
+    
+    /** Stores what functions is being deferred*/
+    MVKDeferredOperationFunctionType functionType;
+    
+    /** The parameters in the operation being deferred*/
+    std::vector<void*> functionParameters = {};
```

Comment:
@billhollings `MVKSmallVector` seems to freak out when using type `void*` because it says `MoltenVK/Utility/MVKSmallVector.h:148:50 Argument may not have 'void' type` and `MoltenVK/Utility/MVKSmallVector.h:305:14 Cannot form a reference to 'void'`.

Reply:
Ah. Okay. Thanks fore reporting that back. That's a limitation of `MVKSmallVector` then. We'd have to upgrade `MVKSmallVector` to accommodate `void*` content. Not sure if it's worth the effort for one use case so far. So I'm okay with using `std::vector`. But perhaps add a comment to the side if it:

```
std::vector<void*> functionParameters = {};    // Can't use MVKSmallVector on void* type.
```

BTW...as you can see elsewhere, we don't generally comment protected member var declarations, so feel free to remove the other comments you have here (we do comment public member vars).

Reply:
Alright then, thanks

Reply:
I am the original contributor of MVKSmallVector.
MVKSmallVector should work just fine with type void *, I just created a sample project and added the line

` MVKSmallVector<void  *,  4>  x  =  {};
`
and got no compiler issues.

So I am not sure why it fails on your side.

Reply:
Something to note is that I'm using the macos 14 beta. I'll try to compile it again later

Reply:
I am using macOS Ventura 13.2 with XCode 14.3.1 and it compiles just fine. When using MVKSmallVector it is recommend to provide a second template argument to specify a pre allocated storage of elements to avoid new/delete's. So use it like this:

`MVKSmallVector<void  *, 4>  x  =  {};`

Some classes in MoltenVK do not provide one, so it's then essentially always allocating storage from the heap. This is fine, unless the vector is allocated/deallocated each frame, in which case the default argument should be used.